packages/phone-number/lib/parse/index.js

```diff
@@ -19,56 +20,49 @@ export default function parse({ input, countryCode = 'US' }) {
     extension: null,
 
   };
-  const cleanInput = (input || '').replace(cleanRegex, '');
-  const startWithPlus = cleanInput[0] === '+';
-  const withoutPlus = cleanInput.replace(plusRegex, '');
-  const startWithStar = withoutPlus[0] === '*';
-  if (startWithPlus && startWithStar) {
+
+  var phoneNumberIsValid = phoneNumber => phoneNumber.length > 3 && validRegExp.test(phoneNumber);
+
+  var isServiceNumber = tokens => !tokens[0] && tokens[1];
+
+  var hasPlus = tokens => tokens[0][0] === '+';
+
+  var phoneNumberLessThanDefine = phoneNumber => phoneNumber.length <= 6;
+
+  var parseCountrycodeAndPhone = () => {
+    const {
+      country = null,
+      phone = null,
+    } = parseNumber(result.phoneNumber, countryCode);
+    result.parsedCountry = country;
+    result.parsedNumber = phone;
+  }
+
+  var formatInput = input => {
+    const inputWithPlus = (input || '').replace(cleanRegex, '');
+    return inputWithPlus[0]+inputWithPlus.slice(1).replace(plusRegex, '');
+  }
+
+  const cleanInput = formatInput(input);
+  const tokens = cleanInput.split(extensionDelimiter);
+
+  if (!phoneNumberIsValid(cleanInput)){
     result.isValid = false;
-  } else {
-    const tokens = withoutPlus.split(extensionDelimiter);
-    if (startWithStar) {
-      if (tokens[1] && tokens[1].length) {
-        result.isServiceNumber = true;
-        result.phoneNumber = `*${tokens[1]}`;
-      } else {
-        result.isValid = false;
-      }
-    } else if (startWithPlus) {
-      if (tokens[0] && tokens[0].length) {
-        result.hasPlus = true;
-        result.phoneNumber = `+${tokens[0]}`;
-        const {
-          country = null,
-          phone = null,
-        } = parseNumber(result.phoneNumber, countryCode);
-        result.parsedCountry = country;
-        result.parsedNumber = phone;
-        if (tokens[1] && tokens[1].length) {
-          result.extension = tokens[1];
-        }
-      } else {
-        result.isValid = false;
-      }
-    } else if (tokens[0] && tokens[0].length) {
-      if (tokens[0].length > 6) {
-        result.phoneNumber = tokens[0];
-        const {
-          country = null,
-          phone = null,
-        } = parseNumber(result.phoneNumber, countryCode);
-        result.parsedCountry = country;
-        result.parsedNumber = phone;
-        if (tokens[1] && tokens[1].length) {
-          result.extension = tokens[1];
-        }
-      } else {
-        result.isExtension = true;
-        result.phoneNumber = tokens[0];
-      }
-    } else {
-      result.isValid = false;
-    }
+    return result;
   }
+
+  if (isServiceNumber(tokens)) {
+    result.isServiceNumber = true;
+    result.phoneNumber = `*${tokens[1]}`;
+    return result;
+  }
+
+  //dont know the relationship between two column bellow
```

Comment:
It is a good practice to refactor code and make the code more readable and easy to understand, but if you don't know something, why don't you raise an issue and get clarification first?